src/libprojectM/PCM.cpp

```diff
@@ -24,381 +24,528 @@
+class AutoLevel
+{
+private:
+    double level;
+    // accumulate sample data
+    size_t level_samples;
+    double level_sum;
+    double level_max;
+    double l0,l1,l2;
+
+public:
+    AutoLevel() : level(0.01),level_samples(0),level_sum(0),level_max(0),l0(-1),l1(-1),l2(-1)
+    {
+    }
 
-  //Initialize buffers to 0
-  for (i=0;i<samples;i++)
+    /*
+     * Here is where we try to do auto volume setting.  Doing this here
+     * means that none of the code downstream (waveforms, beatdetect, etc) needs
+     * to worry about it.
+     *
+     * 1) Don't over react to level changes within a song
+     * 2) Ignore silence/gaps
+     *
+     * I don't know if it's necessary to have both sum and max, but that makes
+     * it easier to experiment...
+     */
+    double updateLevel(size_t samples, double sum, double max)
     {
-      PCMd[0][i]=0;
-      PCMd[1][i]=0;
+        if (sum/samples < 0.00001)
+            return level;
+        level_sum += sum;
+        level_max = fmax(level_max,max*1.02);
+        level_samples += samples;
+        if (level_samples >= 4096 || l0 <= 0)
+        {
+            double max_recent = fmax(fmax(l0, l1), fmax(l2, level_max));
+            l0 = l1; l1 = l2; l2 = level_max; level_max *= 0.95;
+            level_sum = level_samples = 0;
+            level = (l0 <= 0) ? max_recent : level * 0.96 + max_recent * 0.04;
+            level = fmax(level,0.0001);
+        }
+        return level;
     }
+};
+
 
-  start=0;
+PCM::PCM() : start(0), newsamples(0)
+{
+    leveler = new AutoLevel();
 
-  //Allocate FFT workspace
-  // per rdft() documentation
-  //    length of ip >= 2+sqrt(n) and length of w == n/2
+    //Allocate FFT workspace
+    // per rdft() documentation
+    //    length of ip >= 2+sqrt(n) and length of w == n/2
 #if FFT_LENGTH > 1024
 #error update this code
 #endif
-  w  = (double *)wipemalloc(FFT_LENGTH/2*sizeof(double));
-  ip = (int *)wipemalloc(34 * sizeof(int));
-  ip[0]=0;
-
-    /** PCM data */
-//    this->maxsamples = 2048;
-//    this->numsamples = 0;
-//    this->pcmdataL = NULL;
-//    this->pcmdataR = NULL;
+    w  = (double *)wipemalloc(FFT_LENGTH*sizeof(double));
+    ip = (int *)wipemalloc(34 * sizeof(int));
+    ip[0]=0;
+
+    memset(pcmL, 0, sizeof(pcmL));
+    memset(pcmR, 0, sizeof(pcmR));
+    memset(freqL, 0, sizeof(freqL));
+    memset(freqR, 0, sizeof(freqR));
+    memset(spectrumL, 0, sizeof(spectrumL));
+    memset(spectrumR, 0, sizeof(spectrumR));
+}
 
-    /** Allocate PCM data structures */
-    pcmdataL=(float *)wipemalloc(this->maxsamples*sizeof(float));
-    pcmdataR=(float *)wipemalloc(this->maxsamples*sizeof(float));
 
+PCM::~PCM()
+{
+    delete leveler;
+    free(w);
+    free(ip);
 }
 
-PCM::~PCM() {
-
-	free(pcmdataL);
-	free(pcmdataR);
-	free(w);
-	free(ip);
+#include <iostream>
 
-	free(PCMd[0]);
-	free(PCMd[1]);
-	free(PCMd);
 
+void PCM::addPCMfloat(const float *PCMdata, size_t samples)
```

Comment:
This is 1ch right?

Reply:
It would be great if the names were really explicit.  I just didn't want to chase down all the existing usages.
Maybe another PR could make new cleaned up names, and make them all consistent?